src/casp/services/api/data_manager/cellarium_general.py

```diff
@@ -114,7 +121,65 @@ def log_user_activity(self, user_id: int, model_name: str, method: str, cell_cou
         :param cell_count: Number of cells processed in this request
         """
         user_activity = models.UserActivity(
-            user_id=user_id, model_name=model_name, method=method, cell_count=cell_count
+            user_id=user_id,
+            request_id=context[ContextKeys.sentry_trace_id],
+            model_name=model_name,
+            method=method,
+            cell_count=cell_count,
+            event=event,
         )
+
         with self.system_data_db_session_maker.begin() as session:
             session.add(user_activity)
+
+    def get_remaining_quota_for_user(self, user: models.User) -> int:
+        """
+        Get remaining cell processing quota for a specific user (i.e. their quota minus cells processed this week)
+
+        :param user: User object to check quota for
+
+        :return: Remaining cell processing quota
+        """
+        # Return the remaining quota, or 0 if it's negative (which might happen if the user exceeds
+        # their quota, either because we allow them to, or because of any weirdness in the
+        # simultaneous processing of cells)
+        return max(user.cell_quota - self.get_cells_processed_this_week_for_user(user=user), 0)
+
+    def get_cells_processed_this_week_for_user(self, user: models.User) -> int:
+        """
+        Get number of cells processed by a specific user this week
+
+        :param user: User object to check cells processed for
+
+        :return: Number of cells processed this week
+        """
+        # Start of week for a user is based on the day of the week they signed up, so we need to
+        # use the day of the week they signed up to calculate the start of the week
+        current_weekday = datetime.datetime.today().weekday()
+        start_of_the_week_for_user = user.created_at.weekday()
+        days_to_subtract = current_weekday - start_of_the_week_for_user
+        if days_to_subtract < 0:
+            days_to_subtract += 7
+        start_of_week = datetime.datetime.today().replace(
+            hour=0, minute=0, second=0, microsecond=0
+        ) - datetime.timedelta(days=days_to_subtract)
+
+        with self.system_data_db_session_maker() as session:
+            cell_count_sum_query = sa.sql.text(
+                """
+                select
```

Comment:
That sounds reasonable.  I think I'll go the second route because I think the query is just complex enough that it would be a little hard to parse in SQLAlchemy form.

Reply:
Yep, that's a good point. Maybe marking somehow that it's a sqlalchemy text would also bea good idea? I believe this exact query couldn't be executed in pure SQL as it has SQLAlhemy specific string interpolation syntax. Smth like `file_name.sql.sa`? (Not sure about this idea myself, but it'd be great to hear your opinion).